File: 0501~1000/0988. Smallest String Starting From Leaf/main.py
"""
Source: https://rb.gy/oxzda
Date: 2023/8/3
Skill:
Ref:
Runtime: 59 ms, faster than 74.15%
Memory Usage: 18.25 MB, less than 14.15%
Time complexity:
Space complexity:
Constraints:

"""

import math
from typing import List, Tuple, Optional
from collections import defaultdict, deque
from heapq import heapify, heappush, heappop, nsmallest, heappushpop
import bisect
import heapq
import functools


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def smallestFromLeaf(self, root: Optional[TreeNode]) -> str:
        def dfs(node):
            if not node.left and not node.right:
                return [chr(ord("a") + node.val)]
            res = []
            if node.left:
                for string in dfs(node.left):
                    res.append(string + chr(ord("a") + node.val))
            if node.right:
                for string in dfs(node.right):
                    res.append(string + chr(ord("a") + node.val))
            return res

        res = dfs(root)
        return min(res)


if __name__ == "__main__":
    s = Solution()

File: 0501~1000/0988. Smallest String Starting From Leaf/main.cpp
/**
 * Source: is.gd/YDcHOl
 * Date: 2024/4/17
 * Skill:
 * Ref:
 * Runtime: 9 ms, faster than 53.79% of C++ online submissions
 * Memory Usage: 20.24 MB, less than 78.11% of C++ online submissions
 * Time complexity:
 * Space complexity:
 * Constraints:
 *
 */


#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>


using namespace std;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution {
public:
    string smallestFromLeaf(TreeNode* root) {
        string res, cur;

        function<void(TreeNode*)> dfs = [&](TreeNode *node) {
            cur = char('a' + node->val) + cur;
            if (node->left) dfs(node->left);
            if (node->right) dfs(node->right);
            if (!node->left && !node->right) {
                if (res.empty() || res > cur) {
                    res = cur;
                }
            }
            cur.erase(0, 1);
        };

        dfs(root);
        return res;
    }
};